src/rpm/package.rs

```diff
@@ -216,31 +208,94 @@ impl RPMPackage {
+    pub fn verify_digests(&self) -> Result<(), RPMError> {
         let mut header = Vec::<u8>::with_capacity(1024);
         // make sure to not hash any previous signatures in the header
         self.metadata.header.write(&mut header)?;
 
-        let recreated_from_content = Self::create_digests_from_readers(
+        let recreated_from_content = Self::create_legacy_header_digests(
             &mut header.as_slice(),
             SeqCursor::new(&[header.as_slice(), self.content.as_slice()]),
         )?;
 
-        let package_contained = self.metadata.get_digests()?;
-        if recreated_from_content == package_contained {
-            Ok(DigestVerificationOutcome::Match)
-        } else {
-            Ok(DigestVerificationOutcome::Mismatch)
+        let md5 = self
+            .metadata
+            .signature
+            .get_entry_data_as_binary(IndexSignatureTag::RPMSIGTAG_MD5);
+        let sha1 = self
+            .metadata
+            .signature
+            .get_entry_data_as_string(IndexSignatureTag::RPMSIGTAG_SHA1);
+        let sha256 = self
+            .metadata
+            .signature
+            .get_entry_data_as_string(IndexSignatureTag::RPMSIGTAG_SHA256);
+
+        if let Ok(md5) = md5 {
+            if md5 != recreated_from_content.header_and_content_digest {
+                return Err(RPMError::DigestMismatchError);
+            }
         }
+
+        if let Ok(sha1) = sha1 {
+            if sha1 != recreated_from_content.header_digest {
+                return Err(RPMError::DigestMismatchError);
+            }
+        }
+
+        if let Ok(sha256) = sha256 {
+            let sha256_calculated = {
+                let mut hasher = sha2::Sha256::default();
+                hasher.update(self.content.as_slice());
+                hex::encode(hasher.finalize())
+            };
+            if sha256 != sha256_calculated {
+                return Err(RPMError::DigestMismatchError);
+            }
+        }
+
+        let payload_digest_val = self
+            .metadata
+            .header
+            .get_entry_data_as_string_array(IndexTag::RPMTAG_PAYLOADDIGEST);
+        let payload_digest_algo = self
+            .metadata
+            .header
+            .get_entry_data_as_u32(IndexTag::RPMTAG_PAYLOADDIGESTALGO);
+
+        if let (Ok(payload_digest_val), Ok(payload_digest_algo)) =
+            (payload_digest_val, payload_digest_algo)
+        {
+            let mut hasher = match payload_digest_algo {
+                a if a == FileDigestAlgorithm::Sha2_256 as u32 => sha2::Sha256::default(),
+                a => {
+                    return Err(RPMError::UnsupportedFileDigestAlgorithm(
+                        FileDigestAlgorithm::from_u32(a)
+                            .unwrap_or_else(|| panic!("unexpected digest algorithm id {}", a)),
+                    ));
+                } // At the present moment even rpmbuild only supports sha256
+            };
```

Comment:
```suggestion
            let mut hasher = match payload_digest_algo {
                FileDigestAlgorithm::Sha2_256 => sha2::Sha256::default(),
                 // At the present moment even rpmbuild only supports sha256
            };
```

Reply:
And move the error handling into the `TryFrom for FileDigestAlgorithm` implementation.